labs/CryptoProofs/CryptoProofs.md

```diff
@@ -0,0 +1,232 @@
+
+*Hint*: For fastest results, use the `abc` prover.
+
+## 2.3 Collision Detection
+
+In cryptography, a *collision* occurs when two different inputs produce the same output. For some cryptographic functions, such as pseudo-random number generators (PRNGs), it may be desirable to demonstrate an absence of collisions. In other functions, such as cryptographic hash functions, collisions are inevitable, but should be difficult to discover. It is easy in Cryptol to ask the solver to search for collisions. (Though finding a solution may not be possible.)
+
+### Exercise 2.3.1 DES Key Collisions
+
+Use the solver to find two different keys and a plaintext such that both keys encrypt that plaintext to the same ciphertext.
+
+```bash
+labs::CryptoProofs::CryptoProofs> :s prover=yices
+labs::CryptoProofs::CryptoProofs> :sat \(k1,k2,pt) -> k1 != k2 /\ DES.encrypt k1 pt == DES.encrypt k2 pt
+(\(k1, k2, pt) -> k1 != k2 /\ DES.encrypt k1 pt == DES.encrypt k2
+                                                               pt)
+  (0x0000000000000000, 0x0100000000000000, 0x0000000000000000) = True
+(Total Elapsed Time: 1.662s, using Yices)
+```
+
+## 2.4 Equivalence Checking
+
+One of the most powerful uses of Cryptol's theorem proving technology is the ability to show equivalence of two different implementations for all possible inputs. One way to use this is to create two functions:
+
+1. A *specification* that is written for clarity and correctness.
+2. An *implementation* that is optimized and intended for usage.
+
+With a strong specification in hand, a user can aggressively optimize, confident that, given that the equivalence proof passes, their implementation is correct.
+
+Using the Software Analysis Workbench, a powerful framework built on Cryptol, it is even possible to carry out equivalence checks (and all of the proofs from this lab as well) on implementations in languages such as C and Java. (For an introduction to SAW, see other labs in the course.)
+
+### Exercise 2.4.1 DES Equivalent Keys
+
+It's inevitable that there are collisions over the set of all key/plaintext pairs in DES, but it may be surprising that they're easy to find with Cryptol's solver. We now know that the two keys you just found encrypt one particular plaintext to the same ciphertext; more concerning would be if they perform the same transformation on *all* plaintexts. Such keys are called *equivalent* keys.
+
+Attempt to prove that the two keys you just found are equivalent keys. That is, prove that these two keyed DES functions are equivalent for all plaintext inputs. *Hint: Use abc*
+
+### Exercise 2.4.2 DES Parity Bits
+
+Having equivalent keys is often considered a weakness in an a cipher. However, in the case of DES, it turns out that this is a result of a design choice. The lowest bit of each byte of a DES key is actually a *parity* bit that is completely ignored by the cihper itself. The value of the parity bit is such that each byte has an odd number of bits set.
```

Comment:
Could you better describe the parity bit? That onc sentence isn't getting the idea across for me.
Here's a good article to find a definition. You could link to it and quote from it --- https://en.wikipedia.org/wiki/Parity_bit